### ISSUE TITLE
Bug in em++ exception handling in case of multiple inheritance

### ISSUE THREAD

**Issue by zajo** [kripken/emscripten#6449]
The following example should work as is but does not:

```
#include <exception>

struct boost_exception {
  int x; //<== comment-out this line to make it work
};

struct my_error: boost_exception, std::exception { };

int main() {
  try {
    throw my_error();
  } catch( boost_exception & ) {
    try {
      throw;
     } catch( my_error & ) {
     }
  }
  return 0;
}
```

Here is the output when I try to run this program:

```
TypeError: Cannot read property 'rethrown' of undefined
    at ___cxa_rethrow (<REDACTED>/test.js:1610:34)
    at Array.___cxa_rethrow__wrapper (<REDACTED>/test.js:10443:4)
    at Object.dynCall_v (<REDACTED>/test.js:10385:29)
    at invoke_v (<REDACTED>/test.js:2025:24)
    at _main (<REDACTED>/test.js:2263:2)
    at Object.asm._main (<REDACTED>/test.js:10571:21)
    at Object.callMain (<REDACTED>/test.js:10846:30)
    at doRun (<REDACTED>/test.js:10904:60)
    at run (<REDACTED>/test.js:10918:5)
    at Object.<anonymous> (<REDACTED>/test.js:11033:1)
```

I'm running this on Node on MacOS.

`em++ --version` reports:

```
emcc (Emscripten gcc/clang-like replacement) 1.37.36 (commit 2da530aa5e6d6bcaf994ed1b2613a98872a428c8)
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

`node --version` reports:

```
v8.9.4
```

**Comment by kripken** [comment]
Looking into this, it seems that one bug is we don't de-adjust the pointer for a rethrow, which this should fix:
```
diff --git a/src/library.js b/src/library.js
index b1e91ef..722626b 100644
--- a/src/library.js
+++ b/src/library.js
@@ -1257,6 +1257,7 @@ LibraryManager.library = {
   __cxa_rethrow__deps: ['__cxa_end_catch', '$EXCEPTIONS'],
   __cxa_rethrow: function() {
     var ptr = EXCEPTIONS.caught.pop();
+    ptr = EXCEPTIONS.deAdjust(ptr);
     if (!EXCEPTIONS.infos[ptr].rethrown) {
       // Only pop if the corresponding push was through rethrow_primary_exception
       EXCEPTIONS.caught.push(ptr)
```
However, that isn't a complete fix here, as then the exception is not caught, something is still wrong.

**Comment by zajo** [comment]
Thank you for looking into this. I remembered that I had tried to simplify my original example by replacing `std::exception` with a type similar to the other base type, but then I didn't get a crash. I just tried several other tests, hopefully this will help you figure out what is going on.

Using the following types:

```
struct std_exception {
  int x, y;
  std_exception( int x, int y ): x(x), y(y) { }
};

struct boost_exception {
  int z;
  boost_exception( int z ): z(z) { }
};

struct my_error: boost_exception, std_exception {
  my_error(): boost_exception(42), std_exception(1,2) { }
};
```

This works (returns 0):

```
int main() {
  try {
    throw my_error();
  } catch( boost_exception & be ) {
    try {
      if( be.z!=42 )
        return __LINE__;
      throw;
    } catch( std_exception & se ) {
      if( se.x!=1 )
        return __LINE__;
      if( se.y!=2 )
        return __LINE__;
    }
  }
  return 0;
}
```

This also works:

```
int main() {
  try {
    throw my_error();
  } catch( boost_exception & be ) {
    try {
      if( be.z!=42 )
        return __LINE__;
      throw;
    } catch( my_error & me ) {
      if( me.z!=42 )
        return __LINE__;
      if( me.x!=1 )
        return __LINE__;
      if( me.y!=2 )
        return __LINE__;
    }
  }
  return 0;
}
```

But this crashes:

```
int main() {
  try {
    throw my_error();
  } catch( boost_exception & be ) {
    try {
      if( be.z!=42 )
        return __LINE__;
      throw;
    } catch( std_exception & se ) {
      if( se.x!=1 )
        return __LINE__;
      if( se.y!=2 )
        return __LINE__;
      try {
        throw;
      } catch( my_error & me ) {
        if( me.z!=42 )
          return __LINE__;
        if( me.x!=1 )
          return __LINE__;
        if( me.y!=2 )
          return __LINE__;
      }
    }
  }
  return 0;
}
```
Output:
```
exception thrown: TypeError: Cannot read property 'rethrown' of undefined,TypeError: Cannot read property 'rethrown' of undefined
    at ___cxa_rethrow (<REDACTED>/test.js:1611:34)
    at Array.___cxa_rethrow__wrapper (<REDACTED>/test.js:10569:4)
    at Object.dynCall_v (<REDACTED>/test.js:10511:29)
    at invoke_v (<REDACTED>/test.js:2026:24)
    at _main (<REDACTED>/test.js:2317:8)
    at Object.asm._main (<REDACTED>/test.js:10696:21)
    at Object.callMain (<REDACTED>/test.js:10971:30)
    at doRun (<REDACTED>/test.js:11029:60)
    at run (<REDACTED>/test.js:11043:5)
    at Object.<anonymous> (<REDACTED>/test.js:11158:1)
<REDACTED>/test.js:113
      throw ex;
      ^
```

**Comment by kripken** [comment]
Thanks, yeah, the base classes matter here, it turns out. Fixes in #6462 - the second part of the bug was the deadjustment code let to throwing a string instead of an integer (just because of how JS iteration works...).
